Handreiking API Design/Handreiking APi Design.md

```diff
@@ -16,24 +18,31 @@ Dit document heeft een tweeledig doel.
 
 **Motivatie en dominante domeinen**
 
-Bij relaties tussen twee domeinen wil je vanuit elk domein de gerelateerde
-gegevens uit het andere domein kunnen opvragen/inkijken. Eén van de domeinen
-wordt als dominant/primair domein beschouwt. Soms is heel duidelijk welke dat
+Bij relaties tussen twee domeinen wil je vanuit beide domeinen de gerelateerde
+gegevens uit het andere domein kunnen opvragen/inkijken. Daarvoor wordt in elk domein een relatie-entiteit gedefinieerd die in feite de representatie van een entiteit in een ander domein is. Zo'n relatie-entiteit kan vervolgens gebruikt worden om aan andere entiteiten binnen het domein te koppelen. Dit is noodzakelijk omdat de relatie tussen de entiteit in het ene domein en de entiteit in het andere domein slechts op logisch niveau kan worden gelegd. Zo'n relatie wordt nooit op technisch niveau gecreëerd. 
+
+In het voorbeeld hieronder zie je dat in het 'Zaken API' domein de relatie entiteit 'zaakverzoek' is gecreëerd wat in feite een representatie is van de entiteit 'verzoek' uit het 'Klantinteracties API' domein. In het 'Klantinteracties API' domein is juist de relatie-entiteit 'objectverzoek' gecreëerd wat een representatie is van de entiteit 'zaak' uit het 'Zaken API' domein.
+De relatie tussen 'zaakverzoek' en 'verzoek' bestaat alleen op logisch niveau en wordt op technisch niveau niet gerealiseerd. Hetzelfde geldt voor de relatie tussen 'objectverzoek' en 'zaak'.
+
+![Voorbeeld relatie-entiteiten](https://github.com/VNG-Realisatie/API-Kennisbank/blob/master/Handreiking%20API%20Design/Relatie-entiteiten.jpg)
+
+Binnen een relatie tussen domeinen wordt altijd één van de domeinen
+als dominant/primair beschouwt. Soms is heel duidelijk welke dat
 is en is dat goed te beargumenteren, andere keren is het een buikgevoel en
 lastiger te beargumenteren, of het is gewoon een keuze omdat er geen
-natuurlijke dominantie kan aangewezen worden. Voorbeelden hiervan zijn:
+natuurlijke dominantie aangewezen kan worden. Voorbeelden hiervan zijn:
 
 (dominante domein in vet)
 
-* **Zaken** en informatieobjecten
-* **Zaken** en klaninteracties
-* **Besluiten** en informatieobjecten
+* **Zaken** en Documenten
+* **Zaken** en Klantinteracties
+* **Besluiten** en Documenten
 
 **Relatieklassen**
 
-Het feit dat een relatie bestaat tussen objecten in twee verschillende domeinen
-wordt altijd in beide domeinen vastgelegd, met een specifieke resource voor de
-relatieklasse. In ieder domein zijn de relatieklassen als het ware tegenhangers
+Zoals hierboven al aangegeven wordt het feit dat een relatie bestaat tussen objecten in twee verschillende domeinen
+altijd in beide domeinen vastgelegd met een specifieke resource voor de
+relatieklasse (de relatie-entiteit). In ieder domein zijn de relatieklassen als het ware tegenhangers
 van elkaar. Indien er extra relatie-informatie op de relatieklasse bijgehouden
 wordt, dan wordt deze _enkel_ in het dominante domein bijgehouden.
```

Comment:
Betekent dat ook niet dat alleen de klasse met de relatie-eigenschappen benaderbaar is voor comsumers van deze API ? De koppetabel in het andere domein zal dat toch moeten worden bijgehouden door de "eigenaar" van de relatieklasse .

Reply:
Volgens mij was dit wel de conclusie van ons gesprek afgelopen maandag. @michielverhoef, kan jij dit (laten) bevestigen?

Reply:
Dat is inderdaad het geval. De gewone consumer applicatie kan en mag de generieke koppeltabel niet zelf benaderen. Dat mag alleen de eigenaar van die relatieklasse, bijvoorbeeld de ZRC voor een ZaakInformatieobject.

Reply:
Ik heb hier net een commentaar op gegeven volgens mij maar het is weg?
In ieder geval: Ja, dat is inderdaad de conclusie. Overigens is het nu ook al zo dat relatietabellen (koppeltabel met alleen twee verwijzingen) alleen door de eigenaar van de relatie (bijvoorbeeld de ZRC voor relatieklasse ZaakInformatieobject) gebruikt mag worden.
Gewone consumers (vakapplicaties) leggen die relatie dus maar 1 keer, in de ZRC.

Reply:
Correct wat @michielverhoef aangeeft. Het is de verantwoordelijkheid van de API om de koppeltabel bij een andere API bij te werken. Als we dit aan consumers overlaten geeft dat veel te veel kans op fouten.

Overigens heb ik echt een hekel aan deze relaties op meerdere plekken vastleggen. Het is complex, zorgt voor inconsistenties en het maakt (of benadrukt in ieder geval) dat verschillende API's tightly coupled zijn. Maar ja, ik begrijp het probleem dat je vanuit alle perspectieven wil zien wat de relaties zijn en dat je daar data voor nodig hebt bij de resources (en niet fuzzy maar allerlei API's moet gaan bevragen).

Reply:
Ik heb dit gegeven in de tekst proberen te verwoorden.